### PR TITLE
Fix bug in entity batch processing that caused entity state size explosion

### PR DESCRIPTION
**Pull request by sebastianburckhardt**
The current implementation of the entity batching contained a bug which caused entities that experienced a high load to stay in a suspended state for a long time (because the resume message was starved).

As a result, we observed situations in which the internal queue in the entity state grows explosively.

This PR
- fixes the suspected reason causing the long delay of the resume message.
- adds a safety measure that un-suspends any entity if it has been suspended for 10 seconds.
- adds very detailed tracing to make it easier to diagnose issues with entity size and entity batch processing in the future.

### Pull request checklist

* [x] My changes **do not** require documentation changes
* [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
* [x] My changes **do not** require any extra work to be leveraged by OutOfProc SDKs
* [x] My changes **do not** change the version of the WebJobs.Extensions.DurableTask package
* [x] Otherwise: Ensure the EventIds are within the supported range in our existing Windows infrastructure.

### CODE REVIEW

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs:880]: what does `TraceFlag('1')` do? Can we add a small inline comment clarifying this?

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs:986]: nitpick: can we say this more directly? For example: resume entity processing if it has been suspended for at least 10 seconds

File: src/WebJobs.Extensions.DurableTask/EndToEndTraceHelper.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/EndToEndTraceHelper.cs:856]: can we un-do this small whitespace diff?

File: src/WebJobs.Extensions.DurableTask/Listener/TaskEntityShim.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/Listener/TaskEntityShim.cs:357]: stylistic nit: any reason for this is a local function? I agree it's helpful to have it abstracted out as a named method, but I think that this is _just_ long enough to be refactored into a class-level method.

File: src/WebJobs.Extensions.DurableTask/EntityScheduler/MessageSorter.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/EntityScheduler/MessageSorter.cs:66]: I realize this is pre-existing code, so just asking this for my own understanding: can you please clarify what each of terms in the if-statement mean? As in - what is the `SendHorizon`, `reorderWindow`, etc. Just trying to reason about this check. Thanks!

**Review by davidmrdavid — COMMENTED** [review]
Just leaving some questions as an entity-internal newie :) . Most of the PR makes sense, so I think I should be ready to approve shortly after this round.

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs:880]: Sure. TraceFlags are essentially a "log inside the log". They create a string of log characters. This is because it can be tricky to figure out what path was taken through the code.

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs:986]: Sure, I agree.

File: src/WebJobs.Extensions.DurableTask/Listener/TaskEntityShim.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/Listener/TaskEntityShim.cs:357]: I suppose it is a matter of preference. I like local functions because code that belongs together logically stays in close proximity in the source file which helps me to read it. For example, I don't have to define and pass "stopwatch" or "onHostStopping" as arguments, and when you read the function you can directly see where those variables come from.

File: src/WebJobs.Extensions.DurableTask/EntityScheduler/MessageSorter.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/EntityScheduler/MessageSorter.cs:66]: the message sorter is designed to reorder/deduplicate messages. As a first approximation, it works like this:
- on sending side: for each destination, timestamp of last message sent. When sending a message, it includes that timestamp in the "Predecessor" field.
- on receiving side: for each source, timestamp of last message received. If it receives a message specifying "Predecessor" that matches what was recorded as last message, receive it and update the recorded timestamp. If it is smaller discard it as duplicate. If it is larger, put it into a buffer for out-of-order message until it matches.

now, this algorithm is modified so it stops making any reordering or deduplication efforts outside the specified tolerance window (default is 30 minutes). The `SendHorizon` is that window on the sender side, and the `ReceiveHorizon` is that window on the receiver side.

Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/EntityScheduler/MessageSorter.cs:66]: And in order to not spend too much time going through the data structures to clean out irrelevant entries, we use `MinIntervalBetweenCollections`

File: src/WebJobs.Extensions.DurableTask/EndToEndTraceHelper.cs
Comment (sebastianburckhardt) [reply src/WebJobs.Extensions.DurableTask/EndToEndTraceHelper.cs:856]: ok.

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs
Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs]: I like, I can see it being really useful.

File: src/WebJobs.Extensions.DurableTask/Listener/TaskEntityShim.cs
Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/Listener/TaskEntityShim.cs:357]: Fair enough, it is indeed just a matter of preference. I won't block on this.

**Review by davidmrdavid — APPROVED** [review]

File: src/WebJobs.Extensions.DurableTask/EntityScheduler/MessageSorter.cs
Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/EntityScheduler/MessageSorter.cs:66]: I see. That definitely allows me to understand the code better but I think I'm still missing a piece of the puzzle. I'd love to sync with you over the details of this later. For now, the changes make sense to me.

**Comment by sebastianburckhardt** [comment]
I am still not seeing these new trace events (e.g. EventId 233) show up in our telemetry. @davidmrdavid , in your experience, is this long delay normal? I know we had to update the filters that control which EventIds get collected, but I would have expected this to take effect by now.

**Comment by davidmrdavid** [comment]
@sebastianburckhardt: if I recall correctly, we had to make a change _somewhere_ in Antares to allow for this EventId, and antares releases can be rather slow (like how we had to wait for the scale controller for Netherite/MSSQL). Some Antares bits release more often than others, but assuming this is part of the slowest payload, then this delay would be "normal".

@cgillum: I recall you made this change (extending the trace eventIds supported by our platform). Do you know if that change should have released already?

**Pull request merged**
